File: apps/api/src/services/audit.service.ts
import type { Request } from "express";
import type { HydratedDocument } from "mongoose";
import {
  DEFAULT_PAGE_SIZE,
  MAX_PAGE_SIZE,
  type AuditAction,
  type AuditEntityType,
  type AuditLog,
  type PaginatedResult,
} from "@believe-ai/shared";
import type { AuditLogDocument } from "@believe-ai/server";
import { auditRepository } from "../repositories/audit.repository.js";
import { logger } from "../config/logger.js";

function toDto(doc: HydratedDocument<AuditLogDocument>): AuditLog {
  return {
    id: doc._id.toString(),
    userId: doc.userId.toString(),
    action: doc.action as AuditAction,
    entityType: doc.entityType as AuditEntityType,
    entityId: doc.entityId ?? null,
    metadata: (doc.metadata ?? {}) as AuditLog["metadata"],
    ip: doc.ip ?? null,
    userAgent: doc.userAgent ?? null,
    createdAt: doc.createdAt.toISOString(),
  };
}

export interface RecordAuditInput {
  action: AuditAction;
  entityType: AuditEntityType;
  entityId?: string | null;
  metadata?: Record<string, string | number | boolean>;
  /**
   * Explicit actor, for routes that aren't behind authMiddleware and so
   * have no req.userId — currently the OAuth callbacks, where identity
   * arrives via the `state` parameter instead.
   */
  userId?: string;
}

export const auditService = {
  /**
   * Writes an audit entry. Deliberately swallows its own failures: an audit
   * write must never break the user action it's describing. Failures are
   * logged so a silently-broken audit trail is still visible in logs.
   */
  async record(req: Request, input: RecordAuditInput): Promise<void> {
    const userId = input.userId ?? req.userId;
    if (!userId) return;
    try {
      await auditRepository.create({
        userId,
        action: input.action,
        entityType: input.entityType,
        entityId: input.entityId ?? null,
        metadata: input.metadata ?? {},
        ip: req.ip ?? null,
        userAgent: req.get("user-agent") ?? null,
      });
    } catch (err) {
      logger.error({ err, action: input.action }, "Failed to write audit log");
    }
  },

  async list(userId: string, page = 1, limit = DEFAULT_PAGE_SIZE): Promise<PaginatedResult<AuditLog>> {
    const safePage = Math.max(1, page);
    const safeLimit = Math.min(MAX_PAGE_SIZE, Math.max(1, limit));
    const [items, total] = await auditRepository.list(userId, safePage, safeLimit);

    return {
      items: items.map(toDto),
      page: safePage,
      limit: safeLimit,
      total,
      totalPages: Math.max(1, Math.ceil(total / safeLimit)),
    };
  },
};

File: apps/backend/services/audit_service.py
"""Audit trail — reads the caller's own entries; record() is exposed for
future ports of the actions that produce them (see repositories/audit_repository.py)."""

from __future__ import annotations

from typing import Any

from bson import ObjectId

from core.logging import get_logger
from models.audit_log import AuditAction, AuditLog
from repositories import audit_repository
from schemas.audit import AuditLogDto
from schemas.pagination import DEFAULT_PAGE_SIZE, PaginatedResult, safe_limit, safe_page, total_pages

logger = get_logger(__name__)


def _to_dto(doc: AuditLog) -> AuditLogDto:
    return AuditLogDto(
        id=str(doc.id),
        userId=str(doc.userId),
        action=doc.action,
        entityType=doc.entityType,
        entityId=doc.entityId,
        metadata=doc.metadata,
        ip=doc.ip,
        userAgent=doc.userAgent,
        createdAt=doc.createdAt.isoformat(),
    )


async def record(
    user_id: ObjectId,
    action: AuditAction,
    entity_type: str,
    entity_id: str | None = None,
    metadata: dict[str, Any] | None = None,
    ip: str | None = None,
    user_agent: str | None = None,
) -> None:
    """Deliberately swallows its own failures: an audit write must never
    break the user action it's describing (same contract as Node's
    auditService.record)."""
    try:
        await audit_repository.create(user_id, action, entity_type, entity_id, metadata or {}, ip, user_agent)
    except Exception as err:  # noqa: BLE001 — deliberately broad, see docstring
        logger.error("Failed to write audit log for action %s: %s", action, err)


async def list_audit_logs(
    user_id: ObjectId, page: int = 1, limit: int = DEFAULT_PAGE_SIZE
) -> PaginatedResult[AuditLogDto]:
    safe_page_ = safe_page(page)
    safe_limit_ = safe_limit(limit)
    items, total = await audit_repository.list_for_user(user_id, safe_page_, safe_limit_)
    return PaginatedResult[AuditLogDto](
        items=[_to_dto(doc) for doc in items],
        page=safe_page_,
        limit=safe_limit_,
        total=total,
        totalPages=total_pages(total, safe_limit_),
    )
